src/placement/pl_map.c

```diff
@@ -618,57 +698,66 @@ pl_select_leader(daos_obj_id_t oid, uint32_t grp_idx, uint32_t grp_size,
 	int                              replica_idx;
 	int                              i;
 
+	start = grp_idx * grp_size;
 	oc_attr = daos_oclass_attr_find(oid, NULL);
-	if (oc_attr->ca_resil != DAOS_RES_REPL) {
+	if (oc_attr->ca_resil == DAOS_RES_EC) {
 		int tgt_nr = oc_attr->u.ec.e_k + oc_attr->u.ec.e_p;
-		int fail_cnt = 0;
-		int idx = grp_idx * grp_size + tgt_nr - 1;
-		bool parity_rebuilding = false;
-		int leader_shard_idx;
 
-		/* For EC object, elect last shard in the group (must to be
-		 * a parity node) as leader.
+		/* For EC object, the leader candidate order is as following:
+		 * 1. The last parity node if healthy, otherwise
+		 * 2. The prior parity to the former one, and the prior if necessary.
+		 * 3. The first healthy one in the given bit_map.
 		 */
-		shard = pl_get_shard(data, idx);
-		while (shard->po_rebuilding || shard->po_shard == -1 ||
-		       shard->po_target == -1) {
-			idx--;
-			if (shard->po_rebuilding)
-				parity_rebuilding = true;
-
-			fail_cnt++;
-			if (fail_cnt > oc_attr->u.ec.e_p) {
-				D_ERROR(DF_OID" fail_cnt %d, exceed e_p %d, "DF_RC"\n",
-					DP_OID(oid), fail_cnt, oc_attr->u.ec.e_p, DP_RC(-DER_IO));
-				return -DER_IO;
-			}
-
-			shard = pl_get_shard(data, idx);
+
+		for (i = start + tgt_nr - 1; i >= start + oc_attr->u.ec.e_k; i--) {
+			if (bit_map != NIL_BITMAP && isclr(bit_map, i - start))
+				continue;
+
+			shard = pl_get_shard(data, i);
+			if (shard->po_target == -1 || shard->po_shard == -1 || shard->po_rebuilding)
+				continue;
+
+			if (pl_target_in_mbs(shard, mbs))
+				goto found;
 		}
 
-		if (fail_cnt == oc_attr->u.ec.e_p) {
-			/* If parity is rebuilding, let's return DER_STALE,
-			 * so object I/O might refresh the pool map and layout
-			 * until parity rebuilt finish.
-			 */
-			if (parity_rebuilding)
-				return -DER_STALE;
-			else
-				return -DER_IO;
+		/* If we do not know which data shards participate in the transaction,
+		 * let's return DER_STALE, then object I/O might refresh the pool map
+		 * and layout until parity rebuilt finish.
+		 */
+
+		if (bit_map == NIL_BITMAP && mbs == NULL) {
+			D_WARN(DF_OID" all parity shards %d are in rebuilding, retry later.\n",
+				DP_OID(oid), oc_attr->u.ec.e_p);
+			return -DER_STALE;
 		}
 
-		D_ASSERT(shard->po_target != -1);
-		D_ASSERT(shard->po_shard != -1);
-		D_ASSERT(!shard->po_rebuilding);
+		for (i = start; i < start + oc_attr->u.ec.e_k; i++) {
+			if (bit_map != NIL_BITMAP && isclr(bit_map, i - start))
+				continue;
+
+			shard = pl_get_shard(data, i);
+			if (shard->po_target == -1 || shard->po_shard == -1 || shard->po_rebuilding)
+				break;
```

Comment:
Once the failure target count > e_p, it should return EIO, see original fail_cnt check.

Reply:
Yes, after the break, it will return -DER_IO via line 749.

Reply:
ah, right. Sorry for noise.